sys/include/net/nanocoap.h

```diff
@@ -273,6 +279,38 @@ typedef struct {
     void *context;                  /**< ptr to user defined context data   */
 } coap_resource_t;
 
+/**
+ * @brief   CoAP resource request handler context
+ */
+struct _coap_request_ctx {
```

Comment:
Shouldn't we have the order of the members swapped? Otherwise this would break applications that cast the user context, right?

Reply:
Applications can not cast `coap_request_ctx_t *` to whatever they expect. But as of now, they can cast the `void *context` member of that struct to whatever they expect. But hiding the definition of `coap_request_ctx_t` and providing `coap_request_get_context()` would be better IMO.

Reply:
> Applications cannot cast `coap_request_ctx_t *` to whatever they expect

Yes, true. My bad. But then I'm worried this is a big API change for applications out there.

Reply:
I think it is reasonable and beneficial. Users who are registering a function prototype with a `void *context` in their `static const coap_resource_t _resources[]` resource array are going to face a compilation error. And then they should look at the documentation and should be able to update the prototype.

Reply:
> But then I'm worried this is a big API change for applications out there.

There are pending use cases for an API change on this (like knowing which client requested this, esp. w/rt security). This change (that adds an extensible pointer) should be the one we need to compatibly add the others later.